Bluetooth: Controller: Add Ext Adv Parameter Validations

cvinayak opened the pull request:
Add Extended Advertising interval parameter validation.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Add Ext Adv Parameter Validations

Add Extended Advertising interval parameter validation.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -3273,14 +3273,26 @@ static void le_set_ext_adv_param(struct net_buf *buf, struct net_buf **evt)
+	min_interval = sys_get_le24(cmd->prim_min_interval);
+
+	if (IS_ENABLED(CONFIG_BT_CTLR_PARAM_CHECK)) {
+		uint32_t max_interval = sys_get_le24(cmd->prim_max_interval);
```
```suggestion
		const uint32_t max_interval = sys_get_le24(cmd->prim_max_interval);
```

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -3273,14 +3273,26 @@ static void le_set_ext_adv_param(struct net_buf *buf, struct net_buf **evt)
+		uint32_t max_interval = sys_get_le24(cmd->prim_max_interval);
+
+		if ((min_interval > max_interval) ||
+		    (min_interval < 0x0020) ||
+		    (max_interval > 0x4000)) {
```
```suggestion
		    (min_interval < BT_LE_ADV_INTERVAL_MIN) ||
		    (max_interval > BT_LE_ADV_INTERVAL_MAX)) {
```

The `BT_LE_ADV_INTERVAL_MAX` is defined as `0x4000`, but the spec states: 
![image](https://user-images.githubusercontent.com/2520413/165913634-648594b2-9fae-4001-b1bf-695271cb76dc.png)

Shouldn't it be `0xFFFFFF`, or am I missing something?

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -3473,16 +3485,28 @@ static void le_set_per_adv_param(struct net_buf *buf, struct net_buf **evt)
+	max_interval = sys_le16_to_cpu(cmd->max_interval);
+
+	if (IS_ENABLED(CONFIG_BT_CTLR_PARAM_CHECK)) {
+		uint32_t min_interval = sys_le16_to_cpu(cmd->min_interval);
```
```suggestion
		const uint32_t min_interval = sys_le16_to_cpu(cmd->min_interval);
```

cvinayak replied:
done

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -3473,16 +3485,28 @@ static void le_set_per_adv_param(struct net_buf *buf, struct net_buf **evt)
+		uint32_t min_interval = sys_le16_to_cpu(cmd->min_interval);
+
+		if ((min_interval > max_interval) ||
+		    (min_interval < 0x0006) ||
+		    (max_interval > 0x2000)) {
```
Let's add `#define`s in `hci.h` for these as well (or use the ones defined by `gap.h`: 
```c
/** Minimum Periodic Advertising Interval (N * 1.25 ms) */
#define BT_GAP_PER_ADV_MIN_INTERVAL             0x0006
/** Maximum Periodic Advertising Interval (N * 1.25 ms) */
#define BT_GAP_PER_ADV_MAX_INTERVAL             0xFFFF
```

Same question as to why the max is `0x2000` instead of `0xFFFF` as defined by the spec: 
![image](https://user-images.githubusercontent.com/2520413/165914130-1e18b808-ee14-4c8e-b047-7d1978055311.png)

cvinayak committed:
Bluetooth: Add defines for primary and periodic advertising intervals

Add HCI defines for Primary Advertising Interval minimum and
maximum, Periodic Advertising Interval minimum and maximum
values.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Add Kconfig to limit advertising interval max

Add Kconfig to limit the Primary Advertising Interval and
Periodic Advertising Interval maximum supported values in
the Controller implementation.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on include/zephyr/bluetooth/hci.h:
```diff
@@ -1278,6 +1278,9 @@ struct bt_hci_cp_le_set_adv_set_random_addr {
 #define BT_HCI_LE_ADV_PROP_ANON                 BIT(5)
 #define BT_HCI_LE_ADV_PROP_TX_POWER             BIT(6)
 
+#define BT_HCI_LE_PRIM_ADV_INTERVAL_MIN         0x000020
+#define BT_HCI_LE_PRIM_ADV_INTERVAL_MAX         0xFFFFFF
```
Should these contain `EXT` to explicitly verify that these are for extended advertising? 
Legacy primary adv interval is still maximum 0x4000:
![image](https://user-images.githubusercontent.com/2520413/165941168-4b189202-7b11-40a6-810f-c6ca442c532d.png)

cvinayak replied:
Legacy advertising interval are defined here:
https://github.com/zephyrproject-rtos/zephyr/blob/672835730b4ae870a1b92d115342444642f87eb6/include/zephyr/bluetooth/hci.h#L842-L844

Thalley replied:
Right, but to the untrained eye these may be confusing to some, but I guess the location of the `#define` is a good enough hint that they are for extended advertising :)

Thalley commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -512,8 +512,21 @@ config BT_CTLR_ADV_SET
 	  Maximum supported advertising sets.
 
+config BT_CTLR_ADV_INTERVAL_MAX
+	int "Primary Advertising Interval Maximum"
+	depends on BT_BROADCASTER
+	range 32 16384 if BT_LL_SW_SPLIT
+	range 32 16777215
+	default 16384 if BT_LL_SW_SPLIT
+	default 16777215
+	help
+	  Supported Primary Advertising Interval Maximum value.
+	  Bluetooth Core Specification v5.3 permitted range is 0x000020 to
+	  0xFFFFFF in 0.625 ms units.
```
Should we hex `hex` instead of `int` to make it easier to compare with the spec values, e.g. `0x000020` and `0xFFFFFF`? 

Also applies to BT_CTLR_ADV_PERIODIC_INTERVAL_MAX

cvinayak replied:
Done

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -3273,14 +3273,27 @@ static void le_set_ext_adv_param(struct net_buf *buf, struct net_buf **evt)
+		const uint32_t max_interval =
+					sys_get_le24(cmd->prim_max_interval);
+
+		if ((min_interval > max_interval) ||
+		    (min_interval < BT_HCI_LE_PRIM_ADV_INTERVAL_MIN) ||
+		    (max_interval > CONFIG_BT_CTLR_ADV_INTERVAL_MAX)) {
```
Will you please add a comment both here and for periodic as to why we compare `max_interval` with a Kconfig instead of the HCI spec maximum? May not be obvious to others

cvinayak replied:
Done
